slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -237,6 +237,45 @@ void SLAMServiceImpl::ProcessDataOffline() {
     CreateMap();
 }
 
+std::string SLAMServiceImpl::GetNextDataFile() {
+    if (offlineFlag) {
+        if (!b_continue_session) {
+            return "";
+        }
+        if (file_list_offline.size() == 0) {
+            file_list_offline = viam::io::ListFilesInDirectory(path_to_data);
+        }
+        if (file_list_offline.size() == 0) {
+            throw std::runtime_error("no data in data directory");
+        }
+        if (current_file_offline == file_list_offline.size()) {
+            LOG(INFO) << "Finished processing offline images";
```

Comment:
This is a slight behavior change, where we log this line before writing the pbstream. However, I expect we will be removing writing the pbstream at shutdown, once we implement `map_rate_sec`.

Reply:
Yes, we did remove writing a final map at shutdown from orbslam. However, in offline mode, orbslam will write a map immediately once it's processed all data: https://github.com/viamrobotics/slam/blob/392d8bcdc1b122c35d0a5f8c15b2b2a2eb2fd0f3/slam-libraries/viam-orb-slam3/orbslam_server_v1.cc#L600-L602.

Oh, interesting about final map saving for cartographer. That sounds fine to do optimization at shutdown for cartographer. I just suggested we remove the final map saving from orbslam, since it wasn't in the spec and there was a bug we either needed to fix or just remove the code. I wonder how we'll handle final optimization for cartographer when we no longer save maps locally as files, but instead return them for cloud storage as a response to `GetInternalState`. Maybe we'll need to run final optimization in the cloud.

Sounds good, I'll add the log line you suggested.